[Zhang Ziqing] iP

ziqing26 opened the pull request:
# Catobot

> Curiosity killed the cat, but satisfaction brought it back.  - The Daily News (Texas)[(more info)](https://www.wikiwand.com/en/Curiosity_killed_the_cat#/Origin_2)

Catobot helps you organize your mind. It's,

- text-based
- easy to learn
- ~~FAST~~ _SUPER_ FAST to use

All you need to do is,

1. download it from [here](https://github.com/ziqing26/ip/releases/tag/A-Jar).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

## Features:

- [x] Managing tasks
- [x] Managing deadlines
- [ ] Reminders(coming soon)

***
If you Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public static void main(String[] args) {
    new Catobot("./data/Catobot.txt").run();
}
```

Commit:
Add Gradle support

ziqing26 committed:
Implement Greet, Echo, Exit

ziqing26 committed:
Implement Add, List

ziqing26 committed:
Implement Mark as Done

ziqing26 committed:
Implement ToDos, Events, Deadlines

ziqing26 committed:
Test using Automated Text UI Testing

ziqing26 committed:
Handle Errors and refactor code

ziqing26 committed:
Implement Delete

ziqing26 committed:
Implement A-ENUM

ziqing26 committed:
Remove unnecessary comments

ziqing26 committed:
Refactor RequestHandlers and Add Javadocs

ziqing26 committed:
Fix format of Javadocs and import statements

ziqing26 committed:
Implement saving the tasks

ziqing26 committed:
Implement basic local time

ziqing26 committed:
Merge branch 'branch-Level-7'

ziqing26 committed:
Merge branch 'branch-Level-8'

ziqing26 committed:
Add more OOPs and organize into packages

Add Parser, Command and subclasses of Commands. Package catobot
and command.

Organization into packages and addition of OOPs are done in one
commit because they are closely linked to another. Without
organizing Catbot, Storage and Ui into packages, it would be
hard to create the interface Executable and it would be hard to
maintain a desirable OOP.

ziqing26 committed:
Put all classes under catobot package

ziqing26 committed:
Add JUnit tests

ziqing26 committed:
Add Javadoc and add LoadingException

ziqing26 committed:
Put static variable declarations before non-static ones

ziqing26 committed:
Implement Find

ziqing26 committed:
Merge branch 'branch-A-JavaDoc'

ziqing26 committed:
Merge branch 'branch-A-CodingStandard'

ziqing26 committed:
Merge branch 'branch-Level-9'

ziqing26 committed:
Tidy up JavaDoc

ziqing26 committed:
Fix text-ui-test

ziqing26 committed:
Update LocalDateTime in Deadline and Event

ziqing26 committed:
Fix minor styling issues.

cheweejia commented on src/main/java/catobot/command/EventCommand.java:
```diff
@@ -0,0 +1,61 @@
+            throw new InvalidEventException();
+        }
+
+        String[] inputs = content.split("event")[1].trim().split(" /at ");
```
Perhaps it will be good to do the formatting of the String in steps?

cheweejia commented on src/main/java/catobot/command/FindCommand.java:
```diff
@@ -0,0 +1,25 @@
+
+    @Override
+    public void execute(TaskList tasks, Ui ui, Storage storage) throws BotException {
+        String keyword = content.substring(CommandType.FIND.getValue().length()).trim();
```
Maybe it will be better if you can break this down into steps?

cheweejia commented on src/test/java/catobot/CatobotTest.java:
```diff
@@ -0,0 +1,87 @@
+            actual = e.getMessage();
+        }
+        assertEquals(expected, actual);
+        assertEquals(expected, actual);
```
There seems to be a repetition here?

ziqing26 replied:
Yes you are right, thanks for pointing it out.

cheweejia commented on src/main/java/catobot/Storage.java:
```diff
@@ -0,0 +1,112 @@
+                switch (type) {
+                case "D":
+                    Deadline deadline = Deadline.of(description, LocalDateTime.parse(input[3], formatter));
+                    if (isDone == 1) {
```
Perhaps you can abstract this out to a function, since this is used multiple times here?

ziqing26 replied:
Sure!

cheweejia commented on src/main/java/catobot/command/Parser.java:
```diff
@@ -0,0 +1,42 @@
+    public static Command parse(String content) throws BotException {
+        CommandType commandType = CommandType.find(content);
+
+        switch (commandType) {
```
Great use of switch :)

itsyme replied:
I like the use of switch here

cheweejia reviewed (commented):
LGTM! Only require minor adjustments here and there.

ziqing26 committed:
Merge branch 'add-gradle-support'

ziqing26 committed:
Add checkstyle

ziqing26 committed:
Merge branch 'branch-A-CheckStyle'

itsyme commented on src/main/java/catobot/Storage.java:
```diff
@@ -0,0 +1,114 @@
+import catobot.item.Deadline;
+import catobot.item.Event;
+import catobot.item.Task;
+import catobot.item.Todo;
```
I like how the imports are organised.

itsyme commented on src/main/java/catobot/command/Command.java:
```diff
@@ -0,0 +1,34 @@
+/**
+ * Represents a command from the user.
+ */
+public class Command implements Executable {
```
I like the implementation of the Command class.

itsyme commented on src/main/java/catobot/command/DeadlineCommand.java:
```diff
@@ -0,0 +1,60 @@
+        }
+        if (!content.contains("/by") || content.split("/by").length < 2) {
+            throw new InvalidDeadlineException();
+        }
```
Perhaps formatting the String in steps would help it be more readable

itsyme commented on src/main/java/catobot/command/Executable.java:
```diff
@@ -0,0 +1,22 @@
+/**
+ * Represents the interface of an executable object.
+ */
+public interface Executable {
```
I like the idea of having an Executable interface

itsyme commented on src/main/java/catobot/exception/EmptyTaskListException.java:
```diff
@@ -0,0 +1,13 @@
+     * Constructor for EmptyTaskListException.
+     */
+    public EmptyTaskListException() {
+        super("There is no tasks currently! Try to add one!");
```
Perhaps "There are ..." would be more appropriate here

itsyme reviewed (commented):
Looks good! Keep up the good work!

ziqing26 committed:
Implement GUI

ziqing26 committed:
Merge branch 'branch-Level-10'

ziqing26 committed:
Fix nitpicks and add Varargs to TaskList constructor

ziqing26 committed:
Fix typo

yewmeng28 commented on src/main/java/catobot/Storage.java:
```diff
@@ -0,0 +1,113 @@
+        return taskList;
+    }
+
+    private void addToList(Task task, ArrayList<Task> list, int isDone) {
```
Perhaps the type of `isDone` could be a `boolean` instead.

yewmeng28 commented on src/main/java/catobot/command/Command.java:
```diff
@@ -0,0 +1,24 @@
+/**
+ * Represents a command from the user.
+ */
+public class Command implements Executable {
```
Good implementation of the `Command` class and the `Executable` interface!

yewmeng28 commented on src/main/java/catobot/command/ExitCommand.java:
```diff
@@ -0,0 +1,32 @@
+    @Override
+    public String execute(TaskList tasks, Storage storage) throws BotException {
+        try {
+            storage.write(tasks.toStringInDoc());
```
Perhaps the `toStringInDoc()` method could be implemented in the `write` method body, just to make it look nicer here.

ziqing26 replied:
My concern would be that might violate the abstraction barrier since the `toStirngInDoc()` make use of specific fields in the object. 😕

yewmeng28 reviewed (commented):
LGTM! Only some minor tweaks to make the code look nicer :)

ziqing26 committed:
Add assertion feature

Catobot does not indicate when the programmer made a mistake.

Assertion feature helps programmer to debug.

Let's add assertion features to verify assumptions about
Catobot in creation of tasks and storage.

ziqing26 committed:
Command, task and storage classes: improve code quality

Command classes have some common behaviours when parsing input.
Task classes have common date formatter. Storage class has
a bulky method.

The common behaviours across command classes cause code
duplication.

Extracting the common behaviour into functions allows us to
parse input of similar formats.

Let's pull up parsing of content in command classes into
static functions in Parser class and extract out loading
tasks into another method.

Abstraction reduces the redundant code and improves
readability.

ziqing26 committed:
TaskList class: use streams for find

ziqing26 committed:
Merge pull request #1 from ziqing26/branch-A-CodeQuality

Command, task and storage classes: improve code quality

ziqing26 committed:
Merge branch 'master' into branch-A-Streams

ziqing26 committed:
Merge pull request #2 from ziqing26/branch-A-Streams

TaskList class: use streams for find

ziqing26 committed:
Merge branch 'master' into branch-A-Assertions

ziqing26 committed:
Merge pull request #3 from ziqing26/branch-A-Assertions

Add assertion feature

ziqing26 committed:
Fix style issue and add .yml file

ziqing26 committed:
Create main.yml

ziqing26 committed:
Update README.md and fix dialog box height

ziqing26 committed:
Add ascii picture to README.md

ziqing26 committed:
Implement C-Sort Command

Tasks are ordered by the timing when they are added.

A sortable list of tasks is more useful as users want deadlines
and events to be sortable according to the date and time.

Let's add sort method to sort tasks chronologically or
lexicographically according to their types.

ziqing26 committed:
Delete duplicated yml file

pss-coder reviewed (commented):
Hi ZiQing, as part of my tutorial for week 5, I am allocated to PR your code for any code quality issues. But looking so far, besides the couple thoughts I've shared on the comments. It looks great. 

I really liked your implementation of the sort feature in your programme. It also brought me back 2030S vibes. 😆 

Perhaps a suggestion you could consider also, given you have alot of String constants in your code, you could consider creating another java file that stores all your String constants for ease of access and modification. Hope this helps :) 

Couple of nits to fix but, LGTM 👍🏽  😄

pss-coder commented on src/main/java/catobot/Storage.java:
```diff
@@ -0,0 +1,119 @@
+        try {
+            Scanner s = new Scanner(storage);
+
+            while (s.hasNext()) {
+                String rawTask = s.nextLine();
+                String[] input = rawTask.split(" \\| ", 4);
+
+                loadTask(input, taskList);
+            }
```
Could we further abstract this into another method ? seems a couple of things is happening to the load() ?

pss-coder commented on src/main/java/catobot/Storage.java:
```diff
@@ -0,0 +1,119 @@
+        assert !description.equals("") : "description cannot be empty";
+        DateTimeFormatter formatter = Task.DATE_FORMAT_DOC;
+
+        switch (type) {
+        case "D":
+            Deadline deadline = Deadline.of(description, LocalDateTime.parse(input[3], formatter));
+            addToList(deadline, taskList, isDone);
+            break;
+        case "E":
+            Event event = Event.of(description, LocalDateTime.parse(input[3], formatter));
+            addToList(event, taskList, isDone);
+            break;
+        case "T":
+            Todo todo = Todo.of(description);
+            addToList(todo, taskList, isDone);
+            break;
+        default:
+            break;
+        }
```
Great implementation of your method. 
Perhaps you could further abstract your method especially this lines of code ? Since this lines of code is addingTask to list, perhaps you could have something like addTaskToList() ?

ziqing26 replied:
If I abstract them out into `addTaskToList()`, would I still need to pass in the tasks?

pss-coder replied:
I would suppose so, given you are passing a couple of fields here from the codes above.

ziqing26 replied:
i am still a bit confused, isn't `addToList()` serve the function? 😄

pss-coder commented on src/main/java/catobot/command/SortCommand.java:
```diff
@@ -0,0 +1,61 @@
+     * @throws BotException If the command is not valid.
+     */
+    @Override
+    public String execute(TaskList tasks, Storage storage) throws BotException {
+        Predicate<Task> condition;
+        switch (taskType) {
+        case DEADLINE:
+            condition = ((t) -> t instanceof Deadline);
+            break;
+        case EVENT:
+            condition = ((t) -> t instanceof Event);
+            break;
+        case TODO:
+            condition = ((t) -> t instanceof Todo);
+            break;
+        default:
+            throw new InvalidCommandException();
+        }
+
+        return tasks.sort(taskType, condition);
```
From your comments, this method sorts your task. Perhaps the name could be changed to sortTasks() ? or something of similar fashion ?

ziqing26 replied:
it is called execute because it implements the Executable interface

pss-coder replied:
Oh Right, I overlooked the name of the class. haha

ziqing26 committed:
Improve the GUI

ziqing26 committed:
Update Ui.png

ziqing26 committed:
Add User Guide

ziqing26 committed:
Set theme jekyll-theme-cayman

ziqing26 committed:
Config build.gradle for proper jar file name

ziqing26 committed:
Merge branch 'master' of github.com:ziqing26/ip

ziqing26 committed:
TaskList class: fix JavaDoc for sort

ziqing26 committed:
Fix font for text box and JavaDoc styles

ziqing26 committed:
Credit profile photos.

ziqing26 committed:
Fix style in User Guide
